mm2src/coins/eth/eth_swap_v2/eth_taker_swap_v2.rs

```diff
@@ -457,37 +458,111 @@ impl EthCoin {
+            .swap_v2_contracts
+            .as_ref()
+            .map(|contracts| contracts.taker_swap_v2_contract)
+            .ok_or_else(|| {
+                FindPaymentSpendError::Internal("Expected swap_v2_contracts to be Some, but found None".to_string())
+            })?;
+        let approve_func = TAKER_SWAP_V2.function(TAKER_PAYMENT_APPROVE)?;
+        let decoded = decode_contract_call(approve_func, taker_payment.unsigned().data())?;
+        let id = match decoded.first() {
+            Some(Token::FixedBytes(bytes)) => bytes,
+            invalid_token => {
+                return MmError::err(FindPaymentSpendError::InvalidData(format!(
+                    "Expected Token::FixedBytes, got {:?}",
+                    invalid_token
+                )))
+            },
+        };
+        // loop to find maker's spendTakerPayment transaction
         loop {
-            let taker_status = self
-                .payment_status_v2(
-                    taker_swap_v2_contract,
-                    decoded[0].clone(), // id from spendTakerPayment
-                    &TAKER_SWAP_V2,
-                    EthPaymentType::TakerPayments,
-                    TAKER_PAYMENT_STATE_INDEX,
-                )
-                .await?;
-            if taker_status == U256::from(TakerPaymentStateV2::MakerSpent as u8) {
-                return Ok(taker_payment.clone());
-            }
             let now = now_sec();
             if now > wait_until {
-                return MmError::err(WaitForPaymentSpendError::Timeout { wait_until, now });
+                return MmError::err(FindPaymentSpendError::Timeout { wait_until, now });
+            }
+
+            let current_block = match self.current_block().compat().await {
+                Ok(b) => b,
+                Err(e) => {
+                    error!("Error getting block number: {}", e);
+                    Timer::sleep(5.).await;
+                    continue;
+                },
+            };
+
+            // get all logged TakerPaymentSpent events from `from_block` till current block
+            let events = match self
+                .events_from_block(taker_swap_v2_contract, "TakerPaymentSpent", from_block, current_block)
+                .await
+            {
+                Ok(events) => events,
+                Err(e) => {
+                    error!(
+                        "Error getting TakerPaymentSpent events from {} block: {}",
+                        from_block, e
+                    );
+                    Timer::sleep(5.).await;
+                    continue;
+                },
+            };
+
+            // this is how spent event looks like in EtomicSwapTakerV2: event TakerPaymentSpent(bytes32 id, bytes32 secret)
+            let found_event = events.into_iter().find(|event| &event.data.0[..32] == id.as_slice());
+
+            if let Some(event) = found_event {
+                if let Some(tx_hash) = event.transaction_hash {
```

Comment:
will this ever be None? will this be a recoverable state then? otherwise we can terminate early.

Reply:
> will this ever be None?

are you talking about `event.transaction_hash`? This type is from a dependency, we should handle it as it is. The `transaction_hash` could be `None` if the log is emitted by a transaction in a pending state. Once the transaction is included in a mined block, the value should be Some.

> will this be a recoverable state then? 

For TPU V2 we aim to have automatic recover process, if `find_taker_payment_spend_tx` return error then refund process will be started
https://github.com/KomodoPlatform/komodo-defi-framework/blob/c700b59ae5bb564241389cd6d6d930d78be4d948/mm2src/mm2_main/src/lp_swap/taker_swap_v2.rs#L1698-L1716

>  otherwise we can terminate early

Could clarify what do you mean by termination? You want to return error and break loop? 
We should try to find transaction in the loop until time is out

Reply:
> The transaction_hash could be None if the log is emitted by a transaction in a pending state. Once the transaction is included in a mined block, the value should be Some.

aren't we getting events till the current *mined* block? so this tx shouldn't be pending?

> For TPU V2 we aim to have automatic recover process

I just meant we can not do nothing about the fact that tx hash is none.
nothing to do with swap recovery.

how I was thinking (which might be wrong) is that some event types don't have a tx hash which means we supplied a bad event id from the beginning meaning that we can't proceed further. this might not be the case though, gotta read more about this, excuse my eth illiteracy 😂

Reply:
> aren't we getting events till the current mined block? so this tx shouldn't be pending?

yeah, you are right as we are using `from` and `to` block filters for `eth_getLogs` API, tx should be confirmed
```rust
    async fn events_from_block(
        &self,
        swap_contract_address: Address,
        event_name: &str,
        from_block: u64,
        to_block: u64,
    ) -> MmResult<Vec<Log>, FindPaymentSpendError> {
        let contract_event = TAKER_SWAP_V2.event(event_name)?;
        let filter = FilterBuilder::default()
            .topics(Some(vec![contract_event.signature()]), None, None, None)
            .from_block(BlockNumber::Number(from_block.into()))
            .to_block(BlockNumber::Number(to_block.into()))
            .address(vec![swap_contract_address])
            .build();
        let events_logs = self
            .logs(filter)
            .await
            .map_err(|e| FindPaymentSpendError::Transport(e.to_string()))?;
        Ok(events_logs)
    }
```

> some event types don't have a tx hash

I think you are confused a bit. Events/logs themselves don't necessarily contain the transaction hash. Instead, the transaction hash is associated with the **transaction** that **emitted the event**. So Log type from web3 just contains info about tx which emitted this log.

Note about event and log words. **event** in Solidity is a way for a contract to **emit logs** during the execution of a function. 
So they are close words, just events refer to the Solidity construct used in the smart contract to emit logs, while logs refer to the actual data that is recorded in the blockchain when events are emitted.

So using from to block range we are looking for Log which was emitted by spend taker payment transaction.

As for empty tx hash I would like to refer to previous comment https://github.com/KomodoPlatform/komodo-defi-framework/pull/2261#discussion_r1837984109 empty event list or none transaction_hash are not 100% that there is no tx which we are looking for, it could be just blockchain indexation delays or other reasons.

Reply:
> i will read further regarding the truncated (not necessarily empty? right?) event list issue, any resourced regarding the indexaction delay issues and such would be so helpful!

I would say that chain reorganization, network latency, node synchronization lag issues could cause missing information issues. These problems usually temporarily, as I understand. You should wait for more confirmed blocks also try to fetch the info from different nodes.

> regarding an event not having a transaction_hash thought, how would we successfully get the event which has None for the transaction_hash and then try again and all of a sudden we get a Some transaction_hash! is that possible? are these event logs mutable?

Logs are not mutable. Also they are tied to transactions. When a transaction calls a smart contract function that emits an event, this event generates log, which is permanently recorded in the blockchain.

But there’s a nuance.  Lest check doc. According to the documentation https://www.chainnodes.org/docs/ethereum/eth_getLogs, a log from a pending transaction might lack a transaction hash, but when the transaction is confirmed, the log should include it.

Therefore, ideally, when we request a list of logs using the `events_from_block` function with `from_block` and `to_block` filters, it should return only logs from confirmed blocks, which means confirmed transactions. In this case, `event.transaction_hash` should ideally always be `Some`. 
https://github.com/KomodoPlatform/komodo-defi-framework/blob/8d5ed46f1d2e935e270fd2144afc7c400ad4860e/mm2src/coins/eth/eth_swap_v2/eth_taker_swap_v2.rs#L530-L531

We dont need to change from_block. However, if Log has `transaction_hash:None`, it doesn't mean Log doesn't have transaction (actually its not correct by itself, as logs are not owners of txs), it means smth went wrong as **eth_getLogs** API with **fromBlock** and **toBlock** filters will use confirmed blocks.

Reply:
Some additional info https://docs.alchemy.com/docs/deep-dive-into-eth_getlogs#what-are-logs-or-events

Logs and events are used synonymously—smart contracts generate logs by firing off events, so logs provide insights into events that occur within the smart contract. Logs can be found on transaction receipts.

Anytime a transaction is mined, we can see event logs for that transaction by making a request to eth_getLogs and then take actions based off those results. For example, if a purchase is being made using crypto payments, we can use eth_getLogs to see if the sender successfully made the payment before providing the item purchased.

Reply:
> However, if Log has transaction_hash:None, it doesn't mean Log doesn't have transaction, it means smth went wrong as eth_getLogs API with fromBlock and toBlock filters will use confirmed blocks.

im missing u between the lines here so let me repeat that to you and see if i got it correctly. `transaction_hash` **MUST** always be `Some` *eventually*, if it was `None` this is a temporary reporting/mining/etc... thing.

Reply:
> > However, if Log has transaction_hash:None, it doesn't mean Log doesn't have transaction, it means smth went wrong as eth_getLogs API with fromBlock and toBlock filters will use confirmed blocks.
> 
> im missing u between the lines here so let me repeat that to you and see if i got it correctly. `transaction_hash` **MUST** always be `Some` _eventually_, if it was `None` this is a temporary reporting/mining/etc... thing.

Yes, in the context of not using "pending" tag in "eth_getLogs" API, we expect `transaction_hash` always be `Some`. 
The best we can do is to repeat loop cycle until time is out, if None occurred.